[visionOS] Add VideoReceiverEndpointManager class

jyavenard opened the pull request:
#### 9e97ba025925ac0cb733e8babbe6d979dd8beb25
<pre>
[visionOS] Add VideoReceiverEndpointManager class
<a href="https://bugs.webkit.org/show_bug.cgi?id=299500">https://bugs.webkit.org/show_bug.cgi?id=299500</a>
<a href="https://rdar.apple.com/161291402">rdar://161291402</a>

Reviewed by Youenn Fablet.

In preparation for the upcoming AudioVideoRendererRemote/RemoteAudioVideoRendererProxy
we move the handling and tracking of the currently assigned VideoTarget to
its own class.

No change is observable behaviour.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_videoReceiverEndpointManager):
(WebKit::GPUConnectionToWebProcess::videoReceiverEndpointManager):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerVideoTarget const):
* Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.h: Added.
* Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.mm: Renamed from Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerManagerProxyCocoa.mm.
(WebKit::VideoReceiverEndpointManager::VideoReceiverEndpointManager):
(WebKit::m_logger):
(WebKit::VideoReceiverEndpointManager::videoTargetForIdentifier):
(WebKit::VideoReceiverEndpointManager::takeVideoTargetForMediaElementIdentifier):
(WebKit::VideoReceiverEndpointManager::handleVideoReceiverEndpointMessage):
(WebKit::VideoReceiverEndpointManager::handleVideoReceiverSwapEndpointsMessage):
(WebKit::VideoReceiverEndpointManager::logChannel const):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm:
(WebKit::handleVideoReceiverEndpointMessage):
(WebKit::handleVideoReceiverSwapEndpointsMessage):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/300616@main">https://commits.webkit.org/300616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16b6c22bbac5ec3e5ad98f297f9d0cf5d92a1b26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123265 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/42979 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33676 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129973 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6336c8d4-1177-4dfd-bb79-5f42f6f06dc6) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51574 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93697 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/184e2202-efea-44de-908c-5d920ecfbc1d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126218 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/34824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110289 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74325 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4dd4b109-a230-49b2-9c22-caf20dbfb6fc) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73491 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/104538 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28671 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132691 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/50215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102185 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/50591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106510 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25937 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25612 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/50070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/55831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -375,6 +382,9 @@ class GPUConnectionToWebProcess
     Ref<RemoteMediaPlayerManagerProxy> m_remoteMediaPlayerManagerProxy;
+#endif
+#if ENABLE(LINEAR_MEDIA_PLAYER)
+    VideoReceiverEndpointManager m_videoReceiverEndpointManager;
```
Worth lazy allocating?

jyavenard replied:
hmmm this isn't the latest version, it is now lazy allocated (re: changes to LazyUniqueRef.h)

jyavenard replied:
I had a style error (it doesn't handle the #ifdef in the constructor thinking I'm indenting a namespace)

youennf commented on Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.h:
```diff
@@ -0,0 +1,81 @@
+
+private:
+    HashMap<WebCore::VideoReceiverEndpointIdentifier, WebCore::PlatformVideoTarget> m_videoTargetCache;
+    struct VideoRecevierEndpointCacheEntry {
```
```suggestion
    struct VideoReceiverEndpointCacheEntry {
```

youennf commented on Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.h:
```diff
@@ -0,0 +1,81 @@
+    WebCore::PlatformVideoTarget takeVideoTargetForMediaElementIdentifier(WebCore::HTMLMediaElementIdentifier, WebCore::MediaPlayerIdentifier);
+    void handleVideoReceiverEndpointMessage(const VideoReceiverEndpointMessage&);
+    void handleVideoReceiverSwapEndpointsMessage(const VideoReceiverSwapEndpointsMessage&);
+    RefPtr<GPUConnectionToWebProcess> protectedConnection() const { return m_gpuConnection.get(); }
```
Private?

youennf commented on Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.mm:
```diff
@@ -24,43 +24,57 @@
     if (cachedEntry->value.playerIdentifier != playerIdentifier) {
         ALWAYS_LOG(LOGIDENTIFIER, "moving target from player ", cachedEntry->value.playerIdentifier->loggingString(), " to player ", playerIdentifier.loggingString());
-        if (RefPtr mediaPlayer = this->mediaPlayer(cachedEntry->value.playerIdentifier))
+        RefPtr remoteMediaPlayerManagerProxy = protectedConnection()->remoteMediaPlayerManagerProxy();
```
```suggestion
        Ref remoteMediaPlayerManagerProxy = protectedConnection()->remoteMediaPlayerManagerProxy();
```
Ditto elsewhere

youennf replied:
Please make sure to change to `Ref` elsewhere, this allows the reader to not have to check why it should not be nullptr checked.

jyavenard committed:
[visionOS] Add VideoReceiverEndpointManager class
https://bugs.webkit.org/show_bug.cgi?id=299500
rdar://161291402

Reviewed by Youenn Fablet.

In preparation for the upcoming AudioVideoRendererRemote/RemoteAudioVideoRendererProxy
we move the handling and tracking of the currently assigned VideoTarget to
its own class.

No change is observable behaviour.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::m_videoReceiverEndpointManager):
(WebKit::GPUConnectionToWebProcess::videoReceiverEndpointManager):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerManagerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerVideoTarget const):
* Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.h: Added.
* Source/WebKit/GPUProcess/media/cocoa/VideoReceiverEndpointManager.mm: Renamed from Source/WebKit/GPUProcess/media/cocoa/RemoteMediaPlayerManagerProxyCocoa.mm.
(WebKit::VideoReceiverEndpointManager::VideoReceiverEndpointManager):
(WebKit::m_logger):
(WebKit::VideoReceiverEndpointManager::videoTargetForIdentifier):
(WebKit::VideoReceiverEndpointManager::takeVideoTargetForMediaElementIdentifier):
(WebKit::VideoReceiverEndpointManager::handleVideoReceiverEndpointMessage):
(WebKit::VideoReceiverEndpointManager::handleVideoReceiverSwapEndpointsMessage):
(WebKit::VideoReceiverEndpointManager::logChannel const):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm:
(WebKit::handleVideoReceiverEndpointMessage):
(WebKit::handleVideoReceiverSwapEndpointsMessage):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/300616@main
